File: Tests/Main Tests/ExcelAddinVB/IRibbonUI.vb
Imports System.Runtime.CompilerServices
Imports System.Runtime.InteropServices

<ComImport(), Guid("000C03A7-0000-0000-C000-000000000046"), TypeLibType(CShort(&H1040))> _
Public Interface IRibbonUI
    <MethodImpl(MethodImplOptions.InternalCall, MethodCodeType:=MethodCodeType.Runtime), DispId(1)> _
    Sub Invalidate()
    <MethodImpl(MethodImplOptions.InternalCall, MethodCodeType:=MethodCodeType.Runtime), DispId(2)> _
    Sub InvalidateControl(<[In](), MarshalAs(UnmanagedType.BStr)> ByVal ControlID As String)
    <MethodImpl(MethodImplOptions.InternalCall, MethodCodeType:=MethodCodeType.Runtime), DispId(3)> _
    Sub InvalidateControlMso(<[In](), MarshalAs(UnmanagedType.BStr)> ByVal ControlID As String)
    <MethodImpl(MethodImplOptions.InternalCall, MethodCodeType:=MethodCodeType.Runtime), DispId(4)> _
    Sub ActivateTab(<[In](), MarshalAs(UnmanagedType.BStr)> ByVal ControlID As String)
    <MethodImpl(MethodImplOptions.InternalCall, MethodCodeType:=MethodCodeType.Runtime), DispId(5)> _
    Sub ActivateTabMso(<[In](), MarshalAs(UnmanagedType.BStr)> ByVal ControlID As String)
    <MethodImpl(MethodImplOptions.InternalCall, MethodCodeType:=MethodCodeType.Runtime), DispId(6)> _
    Sub ActivateTabQ(<[In](), MarshalAs(UnmanagedType.BStr)> ByVal ControlID As String, <[In](), MarshalAs(UnmanagedType.BStr)> ByVal [Namespace] As String)
End Interface

<ComImport(), Guid("000C0395-0000-0000-C000-000000000046"), TypeLibType(CShort(&H1040))> _
Public Interface IRibbonControl
    <DispId(1)> _
    ReadOnly Property Id() As <MarshalAs(UnmanagedType.BStr)> String
    <DispId(2)> _
    ReadOnly Property Context() As <MarshalAs(UnmanagedType.IDispatch)> Object
    <DispId(3)> _
    ReadOnly Property Tag() As <MarshalAs(UnmanagedType.BStr)> String
End Interface

<ComImport(), TypeLibType(CShort(&H1040)), Guid("000C0396-0000-0000-C000-000000000046")> _
Public Interface IRibbonExtensibility
    <MethodImpl(MethodImplOptions.InternalCall, MethodCodeType:=MethodCodeType.Runtime), DispId(1)> _
    Function GetCustomUI(<[In](), MarshalAs(UnmanagedType.BStr)> ByVal RibbonID As String) As <MarshalAs(UnmanagedType.BStr)> String
End Interface

File: Source/Office/Native/IRibbonUI.cs
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System;
using NetOffice.Attributes;

namespace NetOffice.OfficeApi.Native
{
    /// <summary>
    /// NativeInterface IRibbonUI SupportByVersion Office, 12,14,15,16
    /// The object that is returned by the onLoad procedure specified on the customUI tag.
    /// The object contains methods for invalidating control properties and for refreshing the user interface.
    /// </summary>
    [SupportByVersion("Office", 12,14,15,16)]
	[ComImport, ComVisible(true), Guid("000C03A7-0000-0000-C000-000000000046"), TypeLibType((short) 4160)]
	[EntityType(EntityType.IsNativeInterface)]
	public interface IRibbonUI
	{
        #region Methods

        /// <summary>
        /// SupportByVersion Office, 12,14,15,16
        /// Invalidates the cached values for all of the controls of the Ribbon user interface.
        /// </summary>
        [SupportByVersion("Office", 12,14,15,16)]
		[MethodImpl(MethodImplOptions.InternalCall, MethodCodeType=MethodCodeType.Runtime), DispId(1)]
		void Invalidate();

        /// <summary>
        /// SupportByVersion Office, 12,14,15,16
        /// Invalidates the cached value for a single control on the Ribbon user interface.
        /// </summary>
        /// <param name="ControlID">Specifies the ID of the control that will be invalidated</param>
		[SupportByVersion("Office", 12,14,15,16)]
		[MethodImpl(MethodImplOptions.InternalCall, MethodCodeType=MethodCodeType.Runtime), DispId(2)]
		void InvalidateControl([In, MarshalAs(UnmanagedType.BStr)]string ControlID);

        /// <summary>
        /// SupportByVersion Office, 12,14,15,16
        /// Used to invalidate a built-in control.
        /// </summary>
        /// <param name="ControlID">Specified the identifier of the control that will be invalidated.</param>
		[SupportByVersion("Office", 14,15,16)]
		[MethodImpl(MethodImplOptions.InternalCall, MethodCodeType=MethodCodeType.Runtime), DispId(3)]
		void InvalidateControlMso([In, MarshalAs(UnmanagedType.BStr)]string ControlID);

        /// <summary>
        /// SupportByVersion Office, 12,14,15,16
        /// Activates the specified custom tab.
        /// </summary>
        /// <param name="ControlID">Specifies the identifier of the custom Ribbon tab to be activated</param>
		[SupportByVersion("Office", 14,15,16)]
		[MethodImpl(MethodImplOptions.InternalCall, MethodCodeType=MethodCodeType.Runtime), DispId(4)]
		void ActivateTab([In, MarshalAs(UnmanagedType.BStr)]string ControlID);

        /// <summary>
        /// SupportByVersion Office, 12,14,15,16
        /// Activates the specified built-in tab.
        /// </summary>
        /// <param name="ControlID">Specifies the identifier of the custom Ribbon tab to be activated.</param>
		[SupportByVersion("Office", 14,15,16)]
		[MethodImpl(MethodImplOptions.InternalCall, MethodCodeType=MethodCodeType.Runtime), DispId(5)]
		void ActivateTabMso([In, MarshalAs(UnmanagedType.BStr)]string ControlID);

        /// <summary>
        /// SupportByVersion Office, 12,14,15,16
        /// Activates the specified custom tab on the Microsoft Office Fluent Ribbon UI. Uses the fully qualified name of the tab which includes the identifier and the namespace of the tab.
        /// </summary>
        /// <param name="ControlID">Specifies the identifier of the custom Ribbon tab to be activated</param>
        /// <param name="Namespace">Specifies the namespace of the tab element</param>
		[SupportByVersion("Office", 14,15,16)]
		[MethodImpl(MethodImplOptions.InternalCall, MethodCodeType=MethodCodeType.Runtime), DispId(6)]
		void ActivateTabQ([In, MarshalAs(UnmanagedType.BStr)]string ControlID, [In, MarshalAs(UnmanagedType.BStr)]string Namespace);

		#endregion
	}
}
